parser/src/lib.rs

```diff
@@ -63,845 +55,48 @@ impl StreamOnce for StreamType {
-                                                                  TypedIdent::new(arg.clone())
-                                                              })
-                                                              .collect()),
-                                     input)
-                                })
-                        } else {
-                            Ok((Pattern::Ident(TypedIdent::new(id.clone())),
-                                Consumed::Empty(input)))
-                        }
-                    })
-                })
-                .or(record.map(|fields: Vec<_>| {
-                    let mut types = Vec::new();
-                    let mut patterns = Vec::new();
-                    for (id, field) in fields {
-                        match field {
-                            Ok(name) => types.push((id, name)),
-                            Err(pattern) => patterns.push((id, pattern)),
-                        }
-                    }
-                    Pattern::Record {
-                        typ: self.hole_typ.clone(),
-                        types: types,
-                        fields: patterns,
-                    }
-                }))
-                .map(|p| pos::spanned(span, p))
-                .or(between(token(Token::Open(Delimiter::Paren)),
-                            token(Token::Close(Delimiter::Paren)),
-                            self.pattern()))
-                .parse_stream(input)
-        })
-    }
-
-    fn if_else(&self, input: I) -> ParseResult<SpannedExpr<Id>, I> {
-        let start = input.position().start;
-        (token(Token::If),
-         self.expr(),
-         token(Token::Then),
-         self.expr(),
-         token(Token::Else),
-         self.expr())
-            .map(|(_, b, _, t, _, f)| {
-                pos::spanned(Span {
-                                 start: start,
-                                 end: f.span.end,
-                             },
-                             Expr::IfElse(Box::new(b), Box::new(t), Box::new(f)))
-            })
-            .parse_stream(input)
-    }
-
-    fn let_in(&self, input: I) -> ParseResult<Bindings<Id>, I> {
-        let binding = |t| {
-            (try(optional(self.doc_comment()).skip(token(t))),
-             self.parser(ParserEnv::<I, F>::binding))
-                .map(|(comment, mut bind)| {
-                    bind.comment = comment;
-                    bind
-                })
-        };
-        (as_trait(&mut binding(Token::Let)),
-         many(as_trait(&mut binding(Token::And))),
-         token(Token::In).expected("`in` or an expression in the same column as the `let`"))
-            .map(|(b, bindings, _)| {
-                let mut bindings: Vec<_> = bindings;
-                bindings.insert(0, b);
-                Bindings::LetBindings(bindings)
-            })
-            .parse_stream(input)
-    }
+pub fn parse_expr(symbols: &mut IdentEnv<Ident = Symbol>,
```

Comment:
Should we apply the precedence inside `parse_expr` or as a separate second pass? Make sure the normal parser tests use it anyway to get a few more tests passing.